dbms/src/Storages/Page/V2/PageStorage.cpp

```diff
@@ -1086,13 +1091,42 @@ bool PageStorage::gcImpl(bool not_skip, const WriteLimiterPtr & write_limiter, c
         {
             external_pages_remover(external_pages, live_normal_pages);
         }
+
+        // This instance will accept no more write, and we check whether there is any valid page in non writing page file,
+        // If not, it's very possible that all data have been moved to v3, so we try to roll all writing files to make them can be gced
+        if (no_more_insert)
+        {
+            bool has_normal_non_writing_files = false;
+            for (const auto & page_file : page_files)
+            {
+                if ((page_file.getType() == PageFile::Type::Formal) && (page_file.fileIdLevel() < min_writing_file_id_level))
+                {
+                    has_normal_non_writing_files = true;
+                    break;
+                }
+            }
+            if (!has_normal_non_writing_files)
+            {
+                std::unique_lock lock(write_mutex);
+                LOG_FMT_DEBUG(log, "{} No valid pages in non writing page file. Try to roll all {} writing page files", storage_name, write_files.size());
+                idle_writers.clear();
+                for (auto & write_file : write_files)
+                {
+                    auto writer = checkAndRenewWriter(write_file, {0, 0}, /*parent_path_hint=*/"", nullptr, "", /*force*/ true);
+                    idle_writers.emplace_back(std::move(writer));
+                }
+            }
```

Comment:
There is still one situation that we can not reclaim the disk usage. Assume that
```
write_files: [
pf_2_0 (disk size == 0),
pf_3_0 (disk_size == 64MB),
pf_4_0 (disk_size == 64MB),
pf_5_0 (disk_size == 64MB)]
```
even we roll pf_3_0, pf_4_0, pf_5_0 into new empty files, but the min_writing_file_id_level is still `<2,0>` and can not be GC-ed.

* * *

Maybe when **any page_file of writing_files is not empty**, run `checkAndRenewWriter` with force == true,
and change L422~423 to be 
```
bool is_writable = (force) || (!force && page_file.isValid() && page_file.getType() == PageFile::Type::Formal //
                                                                   && isPageFileSizeFitsWritable(page_file, config));
```

Reply:
It's ok here becuase after roll pf_3_0, pf_4_0, pf_5_0, they will be set legacy regardless of the stats of pf_2_0 and reclaim most of their disk usage which should be good enough.
And if we always roll file when force is true, then we may have the risk to keep rolling many empty writing files.

Reply:
After roll, the `min_writing_file_id_level` is still `<2,0>`, and the pf_3_0, pf_4_0, pf_5_0 won't be set to legacy?
https://github.com/pingcap/tiflash/blob/eb26fc529a2b5ba91837218b142f86fe157c518e/dbms/src/Storages/Page/V2/PageStorage.cpp#L1085-L1087
https://github.com/pingcap/tiflash/blob/eb26fc529a2b5ba91837218b142f86fe157c518e/dbms/src/Storages/Page/V2/PageStorage.cpp#L1336-L1366

Reply:
> And if we always roll file when force is true, then we may have the risk to keep rolling many empty writing files.

I means force to roll when any of the writing file is empty

Reply:
Got it. Fixed.